### PR TITLE
feature: hopefully final steps of the smoosh™️

### PR DESCRIPTION
**Pull request by jstuczyn**
this PR is a bit more extensive than I initially planned because of all the spaghetti that had to be untangled to make it all happen.

the general idea is that now it's possible to run your node simultanously in multiple modes. for example `mixnode` + `entry` or `entry` + `exit`, etc.

the modes do the following:
- `mixnode`: allows the node to handle **forward** sphinx packets
- `entry`: allows the node to accepts websocket connections from incoming clients and to handle **final hop** sphinx packets
- `exit`: allows the node to handle **final hop** sphinx packets as well as starts NR and IPR service providers

furthermore, if node runs with `wireguard` enabled, it will start local `authenticator` service provider which will also implicitly enable final hop packet handling regardless of the underlying mode.

there are also various of other smaller (and bigger) changes:
- upon receiving a forward packet, its delay is now clamped so that one could not dos the node by asking it to delay it for, for example, a year
- as mentioned above, node can now run in multiple modes at the same time,
- `/metrics/mixing` endpoint got deprecated (it still provides the same information, however) and might be removed in future release
- introduced `/metrics/packet-stats` endpoint that provides more extensive information about packets received/sent
- it is now possible to control whether the node should log its statistics to console
- the console logs are also updated with the current packet rates
- all nodes now run the "verloc" protocol and measure every other node on the network
- metrics got revamped and unified:
    - all running metrics are stored in a single `NymNodeMetrics` struct
    - there exists a `MetricsAggregator` that listens to any metrics events that might require additional processing
    - any new metrics event types can be easily added by registering a new handler via `register_handler`. the type must implement `MetricsHandler` trait and use unique variant of `MetricsEvent` enum. for example `GatewaySessionStatsHandler`. it is, however, possible to have opaque handlers, such as the already implemented `LegacyMixingStatsUpdater` and `MixnetMetricsCleaner`
- everything in `mixnode` directory has been removed because there was nothing really left there. the mixing socket listener was unified in `nym-node` and similarly `verloc` was also moved there
- `gateway` directory was similarly reduced in size. now it also creates appropriate tasks as opposed to the whole gateway process. eventually it might also be further stripped, but today is not that day.
- removed the generic parameter on the `GatewayStorage` to simplify all the generics down the stack. it wasn't used anyway

cli:
- added `--modes` argument to specify all node modes with a single command (or env variable). for example: `--modes="mixnode,entry"`. can't be used alongside `--modes`
- extended `--mode` argument to allow specifying it multiple times, for example: `--mode mixnode --mode entry`. can't be used alongside `--mode`

config changes:
- replaced `mode` with `modes` to allow setting the node to run with say, entry + mixnode roles simultanously
- added `maximum_forward_packet_delay` to `mixnet.debug` section
- moved `verloc` section from `mixnode.verloc` to top level
- removed `mixnode` section
- removed `entry_gateway` section to `gateway_tasks`
- moved `authenticator` section from `gateway_tasks` (`entry_gateway`) to `service_providers` (`exit_gateway`)
- renamed `exit_gateway` section to `service_providers`
- moved top level `authenticator` section to `service_providers` so that it'd live alongside NR and IPR
- added general `debug` section
- added `metrics` section

- - -
This change is [Reviewable](https://reviewable.io/reviews/nymtech/nym/5201)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -149,15 +148,14 @@ members = [
...
     "tools/internal/testnet-manager",
-    "tools/internal/testnet-manager/dkg-bypass-contract",
+    "tools/internal/testnet-manager/dkg-bypass-contract", "common/verloc", "tools/internal/mixnet-connectivity-check",
 ]
 
 default-members = [
     "clients/native",
     "clients/socks5",
     "explorer-api",
     "gateway",
```
Comment (octol) [inline Cargo.toml]: should we remove `gateway` as well?

File: common/client-libs/mixnet-client/src/client.rs [hunk 2]
```diff
@@ -61,6 +59,15 @@ pub struct Client {
     config: Config,
 }
 
+impl Clone for Client {
+    fn clone(&self) -> Self {
```
Comment (octol) [inline common/client-libs/mixnet-client/src/client.rs]: The implementation seems to indicate it's not really a clone?

File: gateway/src/node/internal_service_providers.rs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        GatewayError,
+    > {
+        let started_nr = self.network_requester.start_service_provider().await?;
+        let started_ipr = self.ip_router.start_service_provider().await?;
+
+        Ok((started_nr, started_ipr))
+    }
+}
+
+pub struct SPMessageRouterBuilder {
```
Comment (octol) [inline gateway/src/node/internal_service_providers.rs]: The smallest of nits: capitalization. Use `SpMessageRouterBuilder` for consistency ;)

File: nym-node/src/node/http/mod.rs [hunk 4]
```diff
@@ -1,40 +1,66 @@
...
+pub mod helpers;
+pub mod middleware;
+pub mod router;
+pub mod state;
+
+type InnerService = IntoMakeServiceWithConnectInfo<Router, SocketAddr>;
+type ConnectInfoExt = AddExtension<Router, ConnectInfo<SocketAddr>>;
+pub type ServeService = Serve<InnerService, ConnectInfoExt>;
+
+pub struct NymNodeHTTPServer {
```
Comment (octol) [inline nym-node/src/node/http/mod.rs]: nit: `NymNodeHttpServer`

File: common/client-libs/mixnet-client/src/client.rs (hunk 2 above)
Comment (jstuczyn) [reply common/client-libs/mixnet-client/src/client.rs]: good catch. I added it for some intermediate code to pass things around. removed : )

File: gateway/src/node/internal_service_providers.rs (hunk 3 above)
Comment (jstuczyn) [reply gateway/src/node/internal_service_providers.rs]: shouldnt it be `SP` though for "Service Provider"?

**Comment by benedettadavico** [comment]
Testing Completion Date: 14/01/2025

Testing Steps Performed:

Ran a series of tests for nodes focusing on the ability to handle multiple modes, network protocols, and configuration adjustments. Validated wireguard and metrics reporting.

Validation:

[x] Test Passed
Issues Found: None

**Pull request merged**